Mob Abilities

Whoneedspacee opened the pull request:
## About The Pull Request

[Video Showcase](https://www.youtube.com/watch?v=fE9AgxZXbPs)

This overhauls the cooldown action subtype to allow for shared cooldowns and for targeted abilities where you can queue up an ability and then click a target.

It's not a perfect system but it's way better than what is there currently by a longshot, and allows you to for example, add a charging action to any mob, not just simplemobs.

## Why It's Good For The Game

Allows for more easy mobs to be made or even allows for some custom mob creation by admins mid-round as opposed to most of these abilities just being hard-coded procs on the mobs.

## Changelog

:cl: Whoneedspacee
refactor: Refactors megafauna abilities to be mob actions which are grantable to any living mob.
admin: Added a verb under admin events which lets you grant mob actions to a marked living mob.
/:cl:

actions-user committed:
Automatic changelog compile [ci skip]

Whoneedspacee committed:
Revert "Automatic changelog compile [ci skip]"

This reverts commit 897f59fdb268dd5871043c30764ac4f80ad76c7b.

Whoneedspacee committed:
Revert "Revert "Merge remote-tracking branch 'tgstation/master' into mobabilities""

This reverts commit 9a9677c7c2b77fdcbbff93e2e76e7d12136f6069.

Whoneedspacee committed:
Revert "Merge remote-tracking branch 'tgstation/master' into mobabilities"

This reverts commit d462b3a0a440b7ac2bfda7393f2da624d7d73ff1.

Whoneedspacee committed:
Merge remote-tracking branch 'tgstation/master'

Whoneedspacee committed:
Starts reworking mob abilities into Actions

sleep check death now takes a second argument for the mob, if it isn't a mob it just checks for deletion

bubblegums abilities have been mostly changed into actions

cooldown actions no longer need to display their maptext and can instead be red when inactive and white when active

cooldown actions can now have a click intercept for targeting abilities

cooldown actions can now share cooldowns so their cooldown is applied to all other actions with the same shared cooldown

Whoneedspacee committed:
Merge remote-tracking branch 'tgstation/master' into mobabilities

Whoneedspacee committed:
Merge remote-tracking branch 'tgstation/master' into mobabilities

Whoneedspacee committed:
Merge remote-tracking branch 'tgstation/master' into mobabilities

Whoneedspacee committed:
Merge branch 'mobabilities' into origin/mobabilities

Whoneedspacee committed:
Merge branch 'mobabilities' into origin/mobabilities

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
adds proper cooldowns to the charge abilities

fixes up blood warp and adds it to its own file

Whoneedspacee committed:
adds signals for blood warp and charge abilities

shared cooldown actions now have an override time

actions can be chosen to be lit up green when selected

Whoneedspacee committed:
adds range to blood warp

fixes moving during the charge ability

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
compile errors fixed

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
reworks colossus to use mob abilities

adds projectileattack as a mob ability

adds a base type for mob cooldown abilities

bubblegums abilities now have their own variables

blood warp now teleports directly where you click

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
Charge now plays a sound at the source properly

Fixes charge dealing damage to the charger

Adds demonic frost miner abilities as projectile attacks

Creates admin verbs that add mob abilities to a marked mob

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
adds more shared cooldowns for other abilities

changes lots of cooldown times to use the time defines

enables text cooldowns on abilities

adds projectile sound as a variable for projectileattack abilities

reworks blood drunk miner to use datum actions

reworks ash drake to use datum actions

fixes frost miner abilities not having a target

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
removes megafauna using activate to outright bypass cooldowns because it looked weird as well as caused some unintended issues in some areas

lava swoop grant and remove now call their super functions properly

trigger in cooldown actions now returns true if it was successful or in the case of players clicking on it, was successful in being clicked on

Whoneedspacee committed:
go away linter

Whoneedspacee committed:
mob stuff idk idc

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
knockdown time

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
adds the charging ability to spiders

fixes transform_weapon

Whoneedspacee committed:
fixes newline linter

Watermelon914 commented on code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm:
```diff
@@ -13,3 +13,18 @@
 	#define COMPONENT_ATOM_EATEN  (1<<0)
+
+//Mob ability signals
+#define COMSIG_BLOOD_WARP "mob_ability_blood_warp"
+#define COMSIG_STARTED_CHARGE "mob_ability_charge_started"
+#define COMSIG_BUMPED_CHARGE "mob_ability_charge_bumped"
+	#define COMPONENT_OVERRIDE_CHARGE_BUMP (1<<0)
+#define COMSIG_FINISHED_CHARGE "mob_ability_charge_finished"
+#define COMSIG_PROJECTILE_FIRING_STARTED "mob_ability_started_projectile"
+#define COMSIG_PROJECTILE_FIRING_FINISHED "mob_ability_fired_projectile"
+#define COMSIG_SPIRAL_ATTACK_START "mob_spiral_attack_start"
+#define COMSIG_SPIRAL_ATTACK_FINISHED "mob_spiral_attack_finished"
+#define COMSIG_SWOOP_ATTACK_STARTED "mob_swoop_attack_started"
+#define COMSIG_SWOOP_INVULNERABILITY_STARTED "mob_swoop_invulnerability_started"
+#define COMSIG_SWOOP_ATTACK_FINISHED "mob_swoop_attack_finished"
+#define COMSIG_LAVA_ARENA_FAILED "mob_lava_arena_failed"
```
needs documenting

Watermelon914 commented on code/datums/action.dm:
```diff
@@ -673,23 +681,70 @@
-		STOP_PROCESSING(SSfastprocess, src)
-	var/timeleft = max(next_use_time - world.time, 0)
-	if(timeleft == 0)
+/datum/action/cooldown/Trigger(atom/A)
```
use more descriptive variable names

Watermelon914 commented on code/datums/actions/mobs/charge.dm:
```diff
@@ -0,0 +1,228 @@
+		DestroySurroundings()
+		hit_target(source, A, charge_damage)
+
+/datum/action/cooldown/mob_cooldown/charge/proc/hit_target(atom/movable/source, atom/A, damage_dealt)
+	if(SEND_SIGNAL(owner, COMSIG_BUMPED_CHARGE, A) & COMPONENT_OVERRIDE_CHARGE_BUMP)
+		return
+	if(!isliving(A))
+		return
+	var/mob/living/L = A
+	L.visible_message("<span class='danger'>[source] slams into [L]!</span>", "<span class='userdanger'>[source] tramples you into the ground!</span>")
+	source.forceMove(get_turf(L))
+	L.apply_damage(damage_dealt, BRUTE, wound_bonus = CANT_WOUND)
+	playsound(get_turf(L), 'sound/effects/meteorimpact.ogg', 100, TRUE)
+	shake_camera(L, 4, 3)
+	shake_camera(source, 2, 3)
```
use more descriptive variable names for atom and living here

Watermelon914 commented on code/datums/actions/mobs/charge.dm:
```diff
@@ -0,0 +1,228 @@
+			owner.forceMove(place)
+			self_placed = TRUE
+			continue
+		var/mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination/B = new /mob/living/simple_animal/hostile/megafauna/bubblegum/hallucination(place)
+		B.appearance = owner.appearance
+		B.name = "[owner]'s hallucination"
+		B.alpha = 127.5
+		B.move_through_mob = owner
+		B.spawn_blood = spawn_blood
+		INVOKE_ASYNC(src, .proc/do_charge, B, target_atom, delay, past)
```
needs more descriptive of a variable name, avoid one letter variable names like these

Watermelon914 commented on code/datums/actions/mobs/dash.dm:
```diff
@@ -0,0 +1,62 @@
+	var/turf/own_turf = get_turf(owner)
+	if(!QDELETED(dash_target))
+		self_dist_to_target += get_dist(dash_target, own_turf)
+	for(var/turf/open/O in RANGE_TURFS(dash_range, own_turf))
```
rename O to something more descriptive

Watermelon914 reviewed (changes requested):
A lot of variable names are 1 letter, they should ideally have a more descriptive name if you're going to refactor action code.

Since you've already got a lot of 1 letter variable names everywhere, I don't think it'd be fair to ask you to change it all, but please change the ones I've listed below

Whoneedspacee committed:
changed single letter variable to descriptive variables

added comments to signals

tralezab commented:
i really like this decoupling

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:
```diff
@@ -298,14 +179,15 @@ Difficulty: Medium
-	dash_cooldown = 0.8 SECONDS
 
-#undef MINER_DASH_RANGE
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/doom/Initialize()
```
```suggestion
/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/doom/Initialize(mapload)
```

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:
```diff
@@ -51,74 +47,41 @@ Difficulty: Medium
+	/// Transform weapon ability
+	var/datum/action/cooldown/mob_cooldown/transform_weapon/transform_weapon
+
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize()
```
```suggestion
/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize(mapload)
```

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm:
```diff
@@ -51,80 +53,73 @@
+	/// Directional shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/dir_shots/alternating/dir_shots
+
+/mob/living/simple_animal/hostile/megafauna/colossus/Initialize()
```
```suggestion
/mob/living/simple_animal/hostile/megafauna/colossus/Initialize(mapload)
```

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:
```diff
@@ -28,18 +28,49 @@
-		INVOKE_ASYNC(src, /mob/living/simple_animal/hostile/.proc/enter_charge, target)
-	else
-		to_chat(src, span_notice("Your charge is still on cooldown!"))
+/mob/living/simple_animal/hostile/asteroid/lobstrosity/Initialize()
```
```suggestion
/mob/living/simple_animal/hostile/asteroid/lobstrosity/Initialize(mapload)
```

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner.dm:
```diff
@@ -51,74 +47,41 @@ Difficulty: Medium
-
-/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/Initialize(mapload)
+	/// Dash ability
+	var/datum/action/cooldown/mob_cooldown/dash/dash
+	/// Kinetic accelerator ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/kinetic_accelerator/kinetic_accelerator
+	/// Transform weapon ability
+	var/datum/action/cooldown/mob_cooldown/transform_weapon/transform_weapon
```
These need to be handled in destroy

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/colossus.dm:
```diff
@@ -51,80 +53,73 @@
-
-/mob/living/simple_animal/hostile/megafauna/colossus/Initialize(mapload)
+	/// Spiral shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/spiral_shots/spiral_shots
+	/// Random shots ablity
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/random_aoe/random_shots
+	/// Shotgun blast ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/shotgun_blast/shotgun_blast
+	/// Directional shots ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/dir_shots/alternating/dir_shots
```
Same here

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/megafauna/demonic_frost_miner.dm:
```diff
@@ -42,205 +42,76 @@ Difficulty: Extremely Hard
 	var/enraged = FALSE
 	/// If the demonic frost miner is currently transforming to its enraged state
 	var/enraging = FALSE
+	/// Frost orbs ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/rapid_fire/shrapnel/frost_orbs
+	/// Snowball Machine gun Ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/rapid_fire/snowball_machine_gun
+	/// Ice Shotgun Ability
+	var/datum/action/cooldown/mob_cooldown/projectile_attack/shotgun_blast/pattern/ice_shotgun
```
same here

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/giant_spider.dm:
```diff
@@ -191,17 +191,48 @@
-	///Whether or not the tarantula is currently walking on webbing.
+	/// Whether or not the tarantula is currently walking on webbing.
 	var/silk_walking = TRUE
+	/// Charging ability
+	var/datum/action/cooldown/mob_cooldown/charge/charge
```
same here

MrMelbert commented on code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:
```diff
@@ -28,18 +28,49 @@
 	robust_searching = TRUE
 	footstep_type = FOOTSTEP_MOB_CLAW
 	gold_core_spawnable = HOSTILE_SPAWN
+	/// Charging ability
+	var/datum/action/cooldown/mob_cooldown/charge/charge
```
same here

MrMelbert commented on code/datums/actions/mobs/charge.dm:
```diff
@@ -0,0 +1,228 @@
+	SEND_SIGNAL(owner, COMSIG_FINISHED_CHARGE)
+	return TRUE
+
+/datum/action/cooldown/mob_cooldown/charge/proc/on_move(atom/source, atom/new_loc)
+	var/expected_dir = next_move_allowed[source]
```
```suggestion
/datum/action/cooldown/mob_cooldown/charge/proc/on_move(atom/source, atom/new_loc)
	SIGNAL_HANDLER

	var/expected_dir = next_move_allowed[source]
```
Also NONE of your signal procs have signal handlers as far as I can see, you should check them all

Whoneedspacee committed:
adds signal_handler to registered singal procs

adds a basic charge that is less strong than the default bubblegum charge

deletes mob actions upon the mob being destroyed

re-adds deleted maploads

Whoneedspacee commented:
update https://www.youtube.com/watch?v=ceKz0M9C8bE

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into mobabilities

Whoneedspacee committed:
fix linter

Mothblocks commented on code/__DEFINES/cooldowns.dm:
```diff
@@ -95,3 +95,8 @@
 
 #define COOLDOWN_TIMELEFT(cd_source, cd_index) (max(0, cd_source.cd_index - world.time))
+
+//Shared cooldowns for actions
```
Can you put this with the rest of the cooldown keys at the top?

Mothblocks commented on code/__DEFINES/dcs/signals/signals_mob/signals_mob_simple.dm:
```diff
@@ -13,3 +13,30 @@
 #define COMSIG_OOZE_EAT_ATOM "ooze_eat_atom"
 	#define COMPONENT_ATOM_EATEN  (1<<0)
+
+//Mob ability signals /datum/action/cooldown/mob_cooldown
```
Can you move a lot of these to their own file? Already not too happy about things like oozes being in here, now that I see it.

Something like signals_mob_abilities.dm or something.

Mothblocks replied:
Also all of these need to document the proc they're called from, and what arguments are passed.

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+/// Starts a cooldown time, will use default cooldown time if an override is not specified, shares that cooldown with similar abilities
+/datum/action/cooldown/proc/StartCooldown(override_cooldown_time, will_share_cooldowns = TRUE)
+	if(shared_cooldown && will_share_cooldowns)
+		for(var/datum/action/cooldown/C in owner.actions - src)
```
Better name

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
-	next_use_time = world.time + cooldown_time
-	button.maptext = MAPTEXT("<b>[round(cooldown_time/10, 0.1)]</b>")
+/// Starts a cooldown time, will use default cooldown time if an override is not specified, shares that cooldown with similar abilities
+/datum/action/cooldown/proc/StartCooldown(override_cooldown_time, will_share_cooldowns = TRUE)
```
The boolean parameter shouldn't be here, this should just be two separate procs, with one being the process after this that starts the cooldown, and nothing else.

The two places in which you actually use `will_share_cooldowns = FALSE` you would call that directly. `StartCooldown()` itself would run the shared_cooldown stuff, though.

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+		else
+			owner.click_intercept = src
+			to_chat(owner, "<b>You are now prepared to use [src]</b>")
+		for(var/datum/action/cooldown/C in owner.actions)
```
Better name

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+			return InterceptClickOn(owner, null, target)
+		if(owner.click_intercept == src)
+			owner.click_intercept = null
+			to_chat(owner, "<b>[src] is no longer active</b>")
```
```suggestion
			to_chat(owner, "<b>[src] is no longer active.</b>")
```

In an ideal world these are not necessary at all, by the way. The icon itself should have some indication to whether or not its active. If it does, then these do not need to exist.

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+			to_chat(owner, "<b>[src] is no longer active</b>")
+		else
+			owner.click_intercept = src
+			to_chat(owner, "<b>You are now prepared to use [src]</b>")
```
```suggestion
			to_chat(owner, "<b>You are now prepared to use [src].</b>")
```

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+		return TRUE
+	return Activate(owner)
+
+/datum/action/cooldown/proc/InterceptClickOn(mob/living/caller, params, atom/A)
```
Better variable name for `A` (such as `target`)

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+	caller.click_intercept = null
+	return TRUE
+
+/datum/action/cooldown/proc/Activate(atom/target_atom)
```
Document

Mothblocks replied:
Also this can just be `target`, `atom` is already in its typepath anyway.

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+	var/time_left = max(next_use_time - world.time, 0)
+	if(button)
+		if(text_cooldown)
+			button.maptext = "<b>[round(time_left/10, 0.1)]</b>"
```
```suggestion
			button.maptext = MAPTEXT("<b>[round(time_left/10, 0.1)]</b>")
```

Maptext without this looks fairly bad.

Mothblocks commented on code/datums/action.dm:
```diff
@@ -673,23 +681,71 @@
+	if(!owner || time_left == 0)
 		button.maptext = ""
+	if(IsAvailable() && owner.click_intercept == src)
+		button.color = rgb(0,255,0)
```
```suggestion
		button.color = COLOR_GREEN
```

?

Mothblocks reviewed (changes requested):
Going to just assume the other stuff isn't yours, though there's plenty upsetting about them

Whoneedspacee committed:
moves cooldown defines to their own file

startcooldown now has two different procs instead of a boolean for sharing cooldowns or only giving itself a cooldown

cooldown actions now call the parent isavailable

cooldown actions now signal their start and end

actions can now be interrupted by a signal return on the start signal

blood warp is now a main cooldown again

a single charge from a single source no longer hits the same target twice or more

mob cooldowns are now only usable while conscious

fixes frost miners shrapnel projectiles being nearly undodgeable

frost miner no longer shoots while he is enraging

drake no longer shoots meteors every single time

Whoneedspacee committed:
fixes linter

Whoneedspacee committed:
fixes linter for real i swear

Mothblocks commented on code/datums/action.dm:
```diff
@@ -671,25 +679,84 @@
-/datum/action/cooldown/proc/StartCooldown()
-	next_use_time = world.time + cooldown_time
-	button.maptext = MAPTEXT("<b>[round(cooldown_time/10, 0.1)]</b>")
+	return (..() && (next_use_time <= world.time))
```
```suggestion
	return ..() && (next_use_time <= world.time)
```

Mothblocks reviewed (changes requested):
Just one thing but I'll self commit it

Mothblocks committed:
Remove unnecessary parentheses
